### ISSUE TITLE
Allow a publisher to reconnect without disconnecting readers

### ISSUE THREAD

**Issue by aler9** [bluenviron/mediamtx#2214]
## Describe the feature

Blank screen or fallback stream should be routed to readers. Blank screen or fallback streams are checked for codec before being used in place of the original stream.

In case the publisher reconnects, the new stream should be checked:

* if tracks are equal to the one of the original stream, the new stream is routed to readers
* otherwise, readers are resetted

**Comment by rnuck2** [comment]
This feature could be very useful to me. Is its implementation in progress?
At the moment, with the latest release (builded from main branch code), i get alwaysAvailable field not existing.
Thank you.

**Comment by mrasvialpha** [comment]
It would also be good if there is an option to wait until a timeout is reached to check if the publisher reconnects and continue sending frame. Currently all the readers are disconnected immediately.

**Comment by efraimjer** [comment]
any news on this feature?

**Comment by xaionaro** [comment]
Also curious if there are any plans to make it happen anytime soon :)

**Comment by JoLoZ** [comment]
Are there any updates regarding this feature?

**Comment by winkmichael** [comment]
If I understand this correct, this would give the user the ability to have a stream and insert a "No Video.' or "No Input Available" type message into the stream itself?

**Comment by epunter** [comment]
This would be an incredibly good feature, currently the only thing I need to get my gold standard home RTSP server

**Comment by m1k1o** [comment]
I would like to implement this feature, because I am looking for a way how to acheive exactly this behavior for RTMP ingest. While my research I found out that [liquidsoap has already something experimental like this issue](https://www.liquidsoap.info/doc-2.1.2/ffmpeg_cookbook.html#live-switch-between-encoded-content) and there is a way how to acheive it with [gstreamer in rust](https://github.com/centricular/rtmp-slate-fallback).

@aler9 is there any idea / hints from your side how to implement this freatue? Some information that would be beneficial for me to know before I fork this repository and start tinkering with possible solutions? Thanks.

**Comment by TagorePde** [comment]
mistserver , have failover streaming feature.

**Comment by alex-rsk** [comment]
Hello, is this feature already implemented?

**Comment by NextWork123** [comment]
hello,  when that feature gonna implement?

**Comment by devnoname120** [comment]
Sorry for going a bit of a tangent but does anyone here know of a similar tool that has that feature? Either self-hosted or SaaS.

**Comment by fragtion** [comment]
> Sorry for going a bit of a tangent but does anyone here know of a similar tool that has that feature? Either self-hosted or SaaS.

Yes. Try ffreplay. It can do exactly this and pairs well with mediamtx. Would still be nice if mediamtx could do it natively, though, as ffreplay is designed to be used more for a wider use case than just stream fallback

**Comment by devnoname120** [comment]
@fragtion Thanks a lot! I will give it a go. After checking I couldn't find an interesting fork but I'm hopeful that someone will jump in considering that the repository was literally archived today.

https://devnoname120.github.io/useful-forks/?repo=ffplayout/ffplayout

**Comment by devnoname120** [comment]
@fragtion Do you have an example I could start from by any chance? Thanks!

**Comment by hqkirkland** [comment]
I would pay money to have this...

I’ve started to learn GStreamer to try and build an interface which would accomplish this (i.e. send blank stream instead of EOF until new track is published to intermediary receiver) 

…BUT I’m finding that to be very, very difficult:)

**Comment by hqkirkland** [comment]
Also: instead of defining a format in config maybe the stream can just “remember” what the last codec/track format was? My RTSP egress command broadcasts in the same codec every time, regardless of the video type.

After x num of seconds without a connection, the player can “forget” and accept different formats.

**Comment by xaionaro** [comment]
> I would pay money to have this...
> 
> I’ve started to learn GStreamer to try and build an interface which would accomplish this (i.e. send blank stream instead of EOF until new track is published to intermediary receiver)
> 
> …BUT I’m finding that to be very, very difficult:)

Just in case, if you are interested, I implemented this use case in another opensource solution (that I had to write as an alternative to `mediamtx` for my use cases), and I'm happy to support your configuration (no money needed). Feel free to write to my email if you are desperate enough to go with some service that is used by a single user :)

**Comment by yesthatjwz** [comment]
I too want this very much. So much so that I have tried to implement it as a proxy, that reads an HLS stream mediamtx and emits a copy of it with re-generated sequence numbers and inserted "colorbars" segments as needed: [hls-continuity.pl](https://www.dnalounge.com/backstage/src/icecast/hls-continuity.pl)

Problem is, it doesn't work very well, as I described in [this blog post](https://www.jwz.org/blog/2025/06/further-hls-antics/).

This really seems like the sort of thing that would be _so_ much easier to accomplish inside of mediamtx than with an external tool, somewhere inside the path manager, silently substituting a disconnected source with a synthetic short video loop or static file. But that would require someone with a better understanding of mediamtx internals than me.

**Comment by aler9** [comment]
I've thought a little about this feature and i think the implementation can be splitted into two parts:

1. Stream hot replacement - when a new publisher replaces an old one, or when a publisher connects within some seconds from the previous one, and when codecs are the same, the resulting stream should be kept open and the two incoming streams concatenated, with no additional configuration required.
2. Empty frames - in case some seconds have passed from the last frame from a publisher, or in case the "always available" feature is enabled, dummy frames should be inserted periodically, as described at the beginning of the issue. This is the part that requires more effort, since frames are codec dependent.

i think the two parts can be implemented independently and in separate moments.

**Comment by Warr1024** [comment]
With hot swapping, you may need a way to resolve which stream to use if multiple sources are trying to publish to it at the same time; a simple "last one to connect wins and others get disconnected" would be enough for some users (those who know/control the codecs ahead of time) to do their own form of blank frame generation matching their specific needs (though, if there's some other metadata way of setting publisher priority that could be easier to use)

**Comment by aler9** [comment]
> you may need a way to resolve which stream to use if multiple sources are trying to publish to it at the same time

We already have a feature that allows to tune this behavior, it's called "overridePublisher":

```
# Allow another client to disconnect the current publisher and publish in its place.
  overridePublisher: yes
```

**Comment by hqkirkland** [comment]
> This is the part that requires more effort, since frames are codec dependent.

I'd find it acceptable, though maybe some would be very opposed to this, to just repeat the last frame sent by a publisher until your defined timeout has lapsed. That may (?) simplify the implementation.

I've found that some players do this while a reconnect is underway.

**Comment by roger-** [comment]
> > This is the part that requires more effort, since frames are codec dependent.
> 
> I'd find it acceptable, though maybe some would be very opposed to this, to just repeat the last frame sent by a publisher until your defined timeout has lapsed. That may (?) simplify the implementation.
> 
> I've found that some players do this while a reconnect is underway.

Repeating the last frame OR a black frame seem like two reasonable options to support.

**Comment by andreymal** [comment]
> just repeat the last frame

Since not every frame is a keyframe, you'll probably get this:

![Image](https://github.com/user-attachments/assets/a6c56438-f323-4755-b2fd-2c73c782ac42)

**Comment by hqkirkland** [comment]
> > just repeat the last frame
> 
> Since not every frame is a keyframe, you'll probably get this 

Hahaha great point, I don't really want to see frames (or faces) that look like they went through a blender.

Also I want to point out this PR which seems to have a similar intent, with audio:
https://github.com/bluenviron/mediamtx/pull/4832

**Comment by aler9** [comment]
This has been implemented with #5335.

Settings of the final implementation are exactly the same as the ones proposed in the original feature request, which is quite strange, since i expected some twists during the development, while the feature went up exactly as expected.
